### ISSUE TITLE
Function vectorization

### ISSUE THREAD

**Issue by oleg-kachan** [geomstats/geomstats#128]
Ideally, all the functions across all geomstats backends that accept array-like inputs and return array-like outputs should be vectorized (or batched, another term for that) thus containing vectors or arrays along 0-axis.

For example, the shape of an array input containing 1 vector of dimension 3 will be `(1, 3)`, the shape of an array input containing 10 matrices of dimension 3x2 will be `(10, 3, 2)`.

By doing this we will be able to write code in numpy-style avoiding loops. The vectorized code can be parallelized by lower-level routines to run on multiple CPU/GPU cores, while loops are obviously sequential.

Note that, in general, native backend functions are inconsistent in that sense. The same native function like `qr` for example, currently is vectorized in tensorflow, but not in numpy and pytorch. This inconsistency should be addressed by our backend wrappers.
